src/cppcache/integration-test/testThinClientPoolExecuteFunctionThrowsException.cpp

```diff
@@ -0,0 +1,362 @@
+
+      sprintf(buf, "execKey-%d", i);
+      CacheableKeyPtr key = CacheableKey::create(buf);
+      regPtr0->put(key, value);
+    }
+    LOG("Put for execKey's on region complete.");
+
+    LOG("Adding filter");
+    CacheableArrayListPtr arrList = CacheableArrayList::create();
+    for (int i = 100; i < 120; i++) {
+      sprintf(buf, "execKey-%d", i);
+      CacheableKeyPtr key = CacheableKey::create(buf);
+      arrList->push_back(key);
+    }
+
+    CacheableVectorPtr filter = CacheableVector::create();
+    for (int i = 100; i < 120; i++) {
+      sprintf(buf, "execKey-%d", i);
+      CacheableKeyPtr key = CacheableKey::create(buf);
+      filter->push_back(key);
+    }
+    LOG("Adding filter done.");
+
+    CacheablePtr args = CacheableBoolean::create(1);
+
+    ExecutionPtr funcExec = FunctionService::onRegion(regPtr0);
+    ASSERT(funcExec != NULLPTR, "onRegion Returned NULL");
+
+    ResultCollectorPtr collector =
+        funcExec->withArgs(args)->withFilter(filter)->execute(
+            exFuncNameSendException, 15);
+    ASSERT(collector != NULLPTR, "onRegion collector NULL");
+
+    CacheableVectorPtr result = collector->getResult();
+
+    if (result == NULLPTR) {
+      ASSERT(false, "echo String : result is NULL");
+    } else {
+      try {
+        for (int i = 0; i < result->size(); i++) {
+          UserFunctionExecutionExceptionPtr uFEPtr =
+              dynCast<UserFunctionExecutionExceptionPtr>(
+                  result->operator[](i));
+          ASSERT(uFEPtr != NULLPTR, "uFEPtr exception is NULL");
+          LOGINFO("Done casting to uFEPtr");
+          LOGINFO("Read expected uFEPtr exception %s ",
+                  uFEPtr->getMessage()->asChar());
+        }
+      } catch (ClassCastException& ex) {
+        std::string logmsg = "";
+        logmsg += ex.getName();
+        logmsg += ": ";
+        logmsg += ex.getMessage();
+        LOG(logmsg.c_str());
+        ex.printStackTrace();
+        FAIL(
+            "exFuncNameSendException casting to string for bool argument "
+            "exception.");
+      } catch (...) {
+        FAIL(
+            "exFuncNameSendException casting to string for bool argument "
+            "Unknown exception.");
+      }
+    }
+
+    LOG("exFuncNameSendException done for bool argument.");
+
+    collector = funcExec->withArgs(arrList)->withFilter(filter)->execute(
+        exFuncNameSendException, 15);
+    ASSERT(collector != NULLPTR, "onRegion collector for arrList NULL");
+    std::this_thread::sleep_for(std::chrono::seconds(2));
+    
+    try {
+        CacheableVectorPtr fil = CacheableVector::create();
+        fil->push_back(CacheableInt32::create(1));
+        ExecutionPtr exe = FunctionService::onRegion(regPtr0);
+        
+        LOGINFO("Executing the exception test it is expected to throw.");
+        CacheableVectorPtr executeFunctionResult3 =
+        funcExec->withArgs(arrList)->withFilter(filter)->execute("ThinClientRegionExceptionTest", 15)->getResult();
```

Comment:
It will always throw and the fact that it does is success. The reason this is good is that it means that the cleaning up of the worker threads in the executeFunctionSH followed its cleanup path. I don't have a way to directly infer that all the workers were cleaned up directly, but the result of that code path is that when a fatal or unknown exception is thrown by the server, that we cleanup properly. If the exception comes through, then we know the exception that is to be thrown at the end of cleanup happened.